### PR TITLE
integrate rustfmt error messages

### PR DESCRIPTION
**Pull request by mookid**
Re #314 
I finally had some time to look at that one, I just fixed a typo on the way. Thanks for the patch, I love it!

cc @nikodemus

@brotzeit What do you think?

### CODE REVIEW

**Comment by brotzeit** [comment]
I will take a look at it the next days.

@tarsius is working on a solution to making rustic modular https://github.com/brotzeit/rustic/pull/95.
Maybe we can start another try to bring rust-mode and rustic together and reduce the maintenance burden.

File: rust-mode.el [hunk 1]
```diff
@@ -198,6 +198,18 @@ to the function arguments.  When nil, `->' will be indented one level."
   :safe #'booleanp
   :group 'rust-mode)
 
+(defcustom rust-format-show-buffer nil
```
Comment (brotzeit) [inline rust-mode.el]: Why not set the default to t ? I think most users want the buffer the be displayed when there's an error ?

File: rust-mode.el [hunk 2]
```diff
@@ -1455,13 +1467,76 @@ This is written mainly to be used as `end-of-defun-function' for Rust."
...
             (error "Rustfmt failed, see *rustfmt* buffer for details"))))
       (delete-file tmpf))))
 
+;; Since we run rustfmt through stdin we get <stdin> markers in the
+;; output. This replaces them with the buffer name instead.
+(defun rust--format-fix-rustfmt-buffer (buffer-name)
+  (with-current-buffer (get-buffer "*rustfmt*")
+    (goto-char (point-min))
+    (while (re-search-forward "--> <stdin>:")
+      (replace-match (format "--> %s:" buffer-name)))))
```
Comment (brotzeit) [inline rust-mode.el]: I think you can drop the `-->`.

File: rust-mode.el [hunk 3]
```diff
@@ -1455,13 +1467,76 @@ This is written mainly to be used as `end-of-defun-function' for Rust."
...
+(defun rust-goto-format-problem ()
+  "Jumps to problem reported by rustfmt, if any.
+
+In case of multiple problems cycles through them. Displays the
+rustfmt complain in the echo area."
+  (interactive)
+  ;; This uses position in *rustfmt* buffer to know which is the next
+  ;; error to jump to, and source: line in the buffer to figure which
+  ;; buffer it is from.
+  (let ((rustfmt (get-buffer "*rustfmt*")))
```
Comment (brotzeit) [inline rust-mode.el]: Maybe introduce a `defvar` for the buffer name ?

File: rust-mode.el (hunk 1 above)
Comment (mookid) [reply rust-mode.el]: I think that the idea is to show the error in the echo area by default.

Comment (brotzeit) [reply rust-mode.el]: I see.

**Comment by brotzeit** [comment]
FYI, using `call-process-region` in `rust--format-call` doesn't allow us to fix this kind of issue https://github.com/brotzeit/rustic/issues/19. But I guess that's not relevant for most users.

**Comment by mookid** [comment]
> 
> 
> FYI, using `call-process-region` in `rust--format-call` doesn't allow us to fix this kind of issue [brotzeit/rustic#19](https://github.com/brotzeit/rustic/issues/19). But I guess that's not relevant for most users.

Thanks for the pointer. A reference to the fixing commit would also be  useful ;)

**Comment by brotzeit** [comment]
I removed the note about that because I thought you might don't like the solution =)

We would have to add some more logic around a call to `make-process`. [This](https://github.com/brotzeit/rustic/blob/master/rustic-compile.el#L201) would be a solution to the problem.

And when we add this it would be silly not to implement translation of ANSI control sequences through xterm-color(or whatever). So it would end up being more additional code and we decided to keep rust-mode a lightweight package.

**Comment by mookid** [comment]
> FYI, using `call-process-region` in `rust--format-call` doesn't allow us to fix this kind of issue [brotzeit/rustic#19](https://github.com/brotzeit/rustic/issues/19). But I guess that's not relevant for most users.

I can't reproduce that one. Let's keep an eye on it.

**Pull request merged**